File: leetcode/0003-longest-substring-without-repeating-characters/lengthOfLongestSubstring.ts
function lengthOfLongestSubstring(s: string): number {
  const set = new Set<string>();
  let leftIndex = 0;
  let maxLength = 0;

  for (let rightIndex = 0; rightIndex < s.length; rightIndex++) {
    const letter = s[rightIndex]!;

    while (set.has(letter)) {
      set.delete(s[leftIndex]!);
      leftIndex++;
    }

    set.add(letter);
    maxLength = Math.max(maxLength, set.size);
  }

  return maxLength;
}

export default lengthOfLongestSubstring;

File: leetcode/0003-longest-substring-without-repeating-characters/lengthOfLongestSubstring.go
package leetcode

func lengthOfLongestSubstring(s string) int {
	var letterMap = make(map[byte]bool)
	var leftIndex = 0
	var maxLength = 0

	for rightIndex := 0; rightIndex < len(s); rightIndex++ {
		var letter = s[rightIndex]
		for letterMap[letter] {

			delete(letterMap, s[leftIndex])
			leftIndex++
		}

		letterMap[letter] = true
		maxLength = max(maxLength, len(letterMap))
	}

	return maxLength
}

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}
